xarray/core/computation.py

```diff
@@ -1843,36 +1846,84 @@ def where(cond, x, y, keep_attrs=None):
     )
 
 
-def polyval(coord, coeffs, degree_dim="degree"):
+@overload
+def polyval(coord: DataArray, coeffs: DataArray, degree_dim: Hashable) -> DataArray:
+    ...
+
+
+@overload
+def polyval(coord: T_Xarray, coeffs: Dataset, degree_dim: Hashable) -> Dataset:
+    ...
+
+
+@overload
+def polyval(coord: Dataset, coeffs: T_Xarray, degree_dim: Hashable) -> Dataset:
+    ...
+
+
+def polyval(
+    coord: T_Xarray, coeffs: T_Xarray, degree_dim: Hashable = "degree"
+) -> T_Xarray:
     """Evaluate a polynomial at specific values
 
     Parameters
     ----------
-    coord : DataArray
-        The 1D coordinate along which to evaluate the polynomial.
-    coeffs : DataArray
-        Coefficients of the polynomials.
-    degree_dim : str, default: "degree"
+    coord : DataArray or Dataset
+        Values at which to evaluate the polynomial.
+    coeffs : DataArray or Dataset
+        Coefficients of the polynomial.
+    degree_dim : Hashable, default: "degree"
         Name of the polynomial degree dimension in `coeffs`.
 
+    Returns
+    -------
+    DataArray or Dataset
+        Evaluated polynomial.
+
     See Also
     --------
     xarray.DataArray.polyfit
-    numpy.polyval
+    numpy.polynomial.polynomial.polyval
     """
     from .dataarray import DataArray
-    from .missing import get_clean_interp_index
-
-    x = get_clean_interp_index(coord, coord.name, strict=False)
+    from .dataset import Dataset
 
     deg_coord = coeffs[degree_dim]
 
-    lhs = DataArray(
-        np.vander(x, int(deg_coord.max()) + 1),
-        dims=(coord.name, degree_dim),
-        coords={coord.name: coord, degree_dim: np.arange(deg_coord.max() + 1)[::-1]},
+    deg_idx_sorted = np.argsort(deg_coord.values)
+    max_deg = int(deg_coord[deg_idx_sorted[-1]])
+
+    def ensure_numeric(data: DataArray) -> DataArray:
+        if data.dtype.kind in "mM":
+            return DataArray(
+                datetime_to_numeric(
+                    data, offset=np.datetime64("1970-01-01"), datetime_unit="ns"
+                ),
+                dims=data.dims,
+                coords=data.coords,
+                attrs=data.attrs,
+            )
+        return data
+
+    if isinstance(coord, Dataset):
+        coord = coord.map(ensure_numeric)
+    else:
+        coord = ensure_numeric(coord)
+
+    # using Horner's method
+    # https://en.wikipedia.org/wiki/Horner%27s_method
+    res = coeffs.isel({degree_dim: int(deg_idx_sorted[-1])}, drop=True) + zeros_like(
+        coord
     )
-    return (lhs * coeffs).sum(degree_dim)
+    deg_idx = len(deg_coord) - 2
+    for deg in range(max_deg - 1, -1, -1):
+        res *= coord
+        if deg_idx >= 0 and deg == int(deg_coord[deg_idx_sorted[deg_idx]]):
+            # this degrees coefficient is provided, if not assume 0
+            res += coeffs.isel({degree_dim: int(deg_idx_sorted[deg_idx])}, drop=True)
```

Comment:
```suggestion
        if deg_idx >= 0 and deg == int(deg_coord[deg_idx_sorted[deg_idx]]):
            # this degrees coefficient is provided, if not assume 0
            res += coeffs.isel({degree_dim: int(deg_idx_sorted[deg_idx])}, drop=True)
```

I found this bit confusing; a  comment or perhaps sorting by `degree_dim` as suggested above will make it cleaerer.

Reply:
With the `sortby` the code looks much cleaner, thx.